satisfy/src/ConCat/BuildDictionary.hs

```diff
@@ -248,6 +250,81 @@ stringToName str =
 -- When mkUniqueGrimily's argument is negative, we see something like
 -- "Exception: Prelude.chr: bad argument: (-52)". Hence the abs.
 
+-- Transform calls to a function that requires a dictionary into one
+-- another one that also takes a tuple of available locally-bound
+-- dictionaries.  (Note that inScope contains a superset of these
+-- variables, including some that will be unbound in the final output
+-- code.)
+
+extendEvVars :: DVarSet -> Var -> DVarSet
+extendEvVars evVars var =
+  if isEvVar var
+  then extendDVarSet evVars var
+  else evVars
+
+extendEvVarsList :: DVarSet -> [Var] -> DVarSet
+extendEvVarsList evVars vars =
+  extendDVarSetList evVars (filter isEvVar vars)
+
+annotateEvidence :: Id -> Id -> Int -> CoreBind -> CoreM CoreBind
+annotateEvidence fnId fnId' typeArgsCount (NonRec var expr) =
+  do let expr' = annotateExpr fnId fnId' typeArgsCount expr
+     return (NonRec var expr')
+annotateEvidence fnId fnId' typeArgsCount (Rec bindings) =
+  do bindings' <- mapM (\ (var, expr) ->
+                          do let expr' = annotateExpr fnId fnId' typeArgsCount expr
+                             return (var, expr'))
+                       bindings
+     return (Rec bindings')
+
+annotateExpr :: Id -> Id -> Int -> CoreExpr -> CoreExpr
+-- annotateExpr _fnId _fnId' _typeArgsCount expr | pprTrace "annotateExpr" (ppr expr) False = undefined
+annotateExpr fnId fnId' typeArgsCount expr0 =
+  go emptyDVarSet expr0
+  where
+    go _evVars expr@(Var _) = expr
+    go _evVars expr@(Lit _) = expr
+    go evVars expr@(collectArgs -> (Var ((== fnId) -> True), args)) =
+       let (tyArgs, valArgs) = splitAt typeArgsCount args
+       in if length tyArgs < typeArgsCount
+          then pprPanic "unsaturated call to target function" (ppr expr)
+          else
+            let evVarExp = mkCoreTup (map Var (dVarSetElems evVars))
+                valArgs' = map (go evVars) valArgs
+            in mkCoreApps (Var fnId') (tyArgs ++ [Type (exprType evVarExp), evVarExp] ++ valArgs')
+    go evVars (App fn arg) =
+      let fn' = go evVars fn
+          arg' = go evVars arg
+      in App fn' arg'
+    go evVars (Lam var body) =
+      let evVars' = extendEvVars evVars var
+          body' = go evVars' body
+      in Lam var body'
+    go evVars (Let (NonRec var rhs) body) =
+      let rhs' = go evVars rhs
+          body' = go (extendEvVars evVars var) body
+      in Let (NonRec var rhs') body'
+    go evVars (Let (Rec bindings) body) =
+      let bindings' = map (\ (var, expr) -> (var, go evVars expr))
```

Comment:
But this is a recursive `let`. You should recur here with the extended `evVars`.

Reply:
Done.